internal/grpc/services/gateway/storageprovider.go

```diff
@@ -1657,8 +1684,19 @@ func (s *svc) listContainer(ctx context.Context, req *provider.ListContainerRequ
 	}
 	providers = getUniqueProviders(providers)
 
-	resPath := req.Ref.GetPath()
-	if len(providers) == 1 && (utils.IsRelativeReference(req.Ref) || resPath == "" || strings.HasPrefix(resPath, providers[0].ProviderPath)) {
+	p, st := s.getPath(ctx, req.Ref, req.ArbitraryMetadataKeys...)
+	if st.Code != rpc.Code_CODE_OK {
+		return &provider.ListContainerResponse{
+			Status: st,
+		}, nil
+	}
+
+	// check if the path is the root folder
+	if path.Clean(p) == "/" {
```

Comment:
Move this code to `listContainerAcrossProviders`. That method is called when there are more than one provider matching a requested path. And there you don't need to call `getPath` as the requests will always be by path. And you can remove the root folder check as well